grammar.js

```diff
@@ -1,284 +1,391 @@
+ *
+ * @param {Rule} rule
+ * @param {boolean?} trailing_separator - The trailing separator to use.
+ *
+ * @return {SeqRule}
+ *
+ */
+function commaSep1(rule, trailing_separator = false) {
+  // return seq(rule, repeat(seq(',', rule)))
+  return listSeq(rule, ',', trailing_separator);
 }
 
-function commaSep(rule) {
-  return optional(commaSep1(rule))
+/**
+ * Creates a rule to optionally match one or more of the rules separated
+ * by a comma and optionally adds a trailing separator (default is false).
+ *
+ * @param {Rule} rule
+ * @param {boolean?} trailing_separator - The trailing separator to use.
+ *
+ * @return {ChoiceRule}
+ *
+ */
+function commaSep(rule, trailing_separator = false) {
+  return optional(commaSep1(rule, trailing_separator));
 }
 
+/**
+ * This callback should take a rule and places it in between a group (e.g. parentheses).
+ * @callback GroupingCallback
+ * @param {Rule} rule
+ *
+ * @return {SeqRule}
+ */
+
+/**
+ * Creates a rule that matches a binary operator in between two rules.
+ * The two rules are then assigned field names left and right respectively.
+ *
+ * @param {GroupingCallback} grouping
+ * @param {Rule} rule
+ *
+ * @return {ChoiceRule}
+ */
 function opt_grouping(grouping, rule) {
-  return choice(grouping(rule), rule)
+  return choice(grouping(rule), rule);
 }
 
+/**
+ * Creates a rule to match a rule surrounded by parentheses.
+ *
+ * @param {Rule} rule
+ *
+ * @return {SeqRule}
+ */
 function parens(rule) {
-  return seq('(', rule, ')')
+  return seq('(', rule, ')');
 }
 
+/**
+ * Creates a rule to match a rule surrounded by brackets.
+ *
+ * @param {Rule} rule
+ *
+ * @return {SeqRule}
+ */
 function brackets(rule) {
-  return seq('[', rule, ']')
+  return seq('[', rule, ']');
 }
 
+/**
+ * Creates a rule to match a rule surrounded by curly brackets.
+ *
+ * @param {Rule} rule
+ *
+ * @return {SeqRule}
+ */
 function curlies(rule) {
-  return seq('{', rule, '}')
+  return seq('{', rule, '}');
 }
 
+/**
+ * Creates a rule that matches a keyword followed by its "effect",
+ * and then a block.
+ *
+ * @param {String} keyword
+ * @param {Rule} effect
+ * @param {Rule} block
+ *
+ * @return {SeqRule}
+ */
 function bodied_block(keyword, effect, block) {
-  return seq(keyword, effect, field('body', block))
+  return seq(keyword, effect, field('body', block));
 }
 
+/**
+ * Creates a rule that matches a binary operator in between two rules.
+ * The two rules are then assigned field names left and right respectively.
+ *
+ * @param {String} operator
+ * @param {Rule} rule
+ *
+ * @return {SeqRule}
+ */
 function binary_operator(operator, rule) {
-  return seq(field('left', rule), operator, field('right', rule))
+  return seq(field('left', rule), operator, field('right', rule));
 }
 
-const identifier_pattern = /[a-zA-Z_]\w*/;
+module.exports = grammar({
+  name: 'openscad',
 
-const rules = {
-  source_file: $ => repeat(choice($.use_statement, $._item)),
+  word: ($) => $.identifier,
```

Comment:
I dislike the parentheses around the `$` here. It goes against the official tree-sitter examples, the repos owned by the tree-sitter github org, and it adds unnecessary visual noise.

Reply:
Yes, I forgot to disable that sorry